File: examples/py/05_output_access.py
"""Example 05: Output Access

Demonstrates accessing outputs in Python. Note: Python bindings primarily
expose the context (final outputs). This example shows branching and data flow.
"""

import sys
sys.path.insert(0, '/home/runner/work/graph-sp/graph-sp/examples/py')

from benchmark_utils import Benchmark, print_header, print_section
import dagex
import time


def source(_inputs):
    """Generate initial data."""
    return {"data": 100}


def processor_a(inputs):
    """Processor A: multiply by 2."""
    value = inputs.get("input", 0)
    
    # Simulate I/O or blocking operation that releases the GIL
    # This allows true parallel execution in Python
    time.sleep(0.02)
    
    return {"processed": value * 2}


def processor_b(inputs):
    """Processor B: add 50."""
    value = inputs.get("input", 0)
    
    # Simulate I/O or blocking operation that releases the GIL
    # This allows true parallel execution in Python
    time.sleep(0.02)
    
    return {"processed": value + 50}


def combine(inputs):
    """Combine outputs from branches."""
    a = inputs.get("a", 0)
    b = inputs.get("b", 0)
    return {"final": a + b + 1}


def main():
    print_header("Example 05: Output Access")
    
    print("📖 Story:")
    print("   Sometimes you need access to intermediate results, not just the")
    print("   final outputs. While Python bindings primarily expose the context")
    print("   (final outputs), this example shows how to work with branching")
    print("   and understand the flow of data through the graph.\n")
    
    print_section("Building the Graph")
    
    graph = dagex.Graph()
    
    # Add source
    graph.add(
        source,
        label="Source",
        inputs=None,
        outputs=[("data", "input")]
    )
    
    # Create branch A
    branch_a = dagex.Graph()
    branch_a.add(
        processor_a,
        label="ProcessorA",
        inputs=[("input", "input")],
        outputs=[("processed", "a")]
    )
    branch_a_id = graph.branch(branch_a)
    
    # Create branch B
    branch_b = dagex.Graph()
    branch_b.add(
        processor_b,
        label="ProcessorB",
        inputs=[("input", "input")],
        outputs=[("processed", "b")]
    )
    branch_b_id = graph.branch(branch_b)
    
    # Add combine node
    graph.add(
        combine,
        label="Combine",
        inputs=[("a", "a"), ("b", "b")],
        outputs=[("final", "output")]
    )
    
    dag = graph.build()
    
    print_section("Mermaid Diagram")
    print(dag.to_mermaid())
    
    print_section("Sequential Execution (parallel=False)")
    
    with Benchmark("Sequential execution") as bench_seq:
        context_seq = dag.execute(parallel=False)
    
    bench_seq.print_result()
    result_seq = bench_seq.result
    
    print_section("Parallel Execution (parallel=True)")
    
    with Benchmark("Parallel execution") as bench_par:
        context_par = dag.execute(parallel=True, max_threads=4)
    
    bench_par.print_result()
    result_par = bench_par.result
    
    print_section("Results")
    
    print("📊 Accessing outputs:\n")
    
    print("Sequential execution:")
    print(f"  Time: {result_seq.duration_ms:.3f}ms")
    
    print("\nParallel execution:")
    print(f"  Time: {result_par.duration_ms:.3f}ms")
    
    # Final context outputs
    print("\nFinal context outputs:")
    output = context_par.get("output")
    if output is not None:
        print(f"   output: {output}")
    
    print("\nExecution flow:")
    print("   Source: 100")
    print("   ProcessorA (branch A): 100 × 2 = 200")
    print("   ProcessorB (branch B): 100 + 50 = 150")
    print("   Combine: 200 + 150 + 1 = 351")
    
    print("\n✅ Successfully accessed outputs!")
    
    print()


if __name__ == "__main__":
    main()

File: examples/rs/05_output_access.rs
// Example 05: Output Access
// Demonstrates accessing individual node and branch outputs

mod benchmark_utils;

use dagex::{Graph, GraphData};
use std::collections::HashMap;
use std::thread;
use std::time::Duration;
use benchmark_utils::{Benchmark, print_header, print_section};

fn source(_inputs: &HashMap<String, GraphData>) -> HashMap<String, GraphData> {
    let mut outputs = HashMap::new();
    outputs.insert("data".to_string(), GraphData::int(100));
    outputs
}

fn processor_a(inputs: &HashMap<String, GraphData>) -> HashMap<String, GraphData> {
    let value = inputs.get("input").and_then(|d| d.as_int()).unwrap_or(0);
    
    // Simulate I/O-bound work (file read, network call, database query, etc.)
    // that benefits from parallelization
    thread::sleep(Duration::from_millis(150));
    
    let mut outputs = HashMap::new();
    outputs.insert("processed".to_string(), GraphData::int(value * 2));
    outputs
}

fn processor_b(inputs: &HashMap<String, GraphData>) -> HashMap<String, GraphData> {
    let value = inputs.get("input").and_then(|d| d.as_int()).unwrap_or(0);
    
    // Simulate I/O-bound work (file read, network call, database query, etc.)
    // that benefits from parallelization
    thread::sleep(Duration::from_millis(150));
    
    let mut outputs = HashMap::new();
    outputs.insert("processed".to_string(), GraphData::int(value + 50));
    outputs
}

fn main() {
    print_header("Example 05: Output Access");
    
    println!("📖 Story:");
    println!("   Sometimes you need access to intermediate results, not just the");
    println!("   final outputs. The execute_detailed() method returns an ExecutionResult");
    println!("   with context (final outputs), node_outputs (per-node results), and");
    println!("   branch_outputs (per-branch results).\n");
    
    print_section("Building the Graph");
    
    let mut graph = Graph::new();
    
    // Add source
    graph.add(
        source,
        Some("Source"),
        None,
        Some(vec![("data", "input")])
    );
    
    // Create branch A
    let mut branch_a = Graph::new();
    branch_a.add(
        processor_a,
        Some("ProcessorA"),
        Some(vec![("input", "input")]),
        Some(vec![("processed", "result_a")])
    );
    let branch_a_id = graph.branch(branch_a);
    
    // Create branch B
    let mut branch_b = Graph::new();
    branch_b.add(
        processor_b,
        Some("ProcessorB"),
        Some(vec![("input", "input")]),
        Some(vec![("processed", "result_b")])
    );
    let branch_b_id = graph.branch(branch_b);
    
    // Merge branches
    graph.merge(
        |inputs: &HashMap<String, GraphData>| -> HashMap<String, GraphData> {
            let a = inputs.get("from_a").and_then(|d| d.as_int()).unwrap_or(0);
            let b = inputs.get("from_b").and_then(|d| d.as_int()).unwrap_or(0);
            let mut outputs = HashMap::new();
            outputs.insert("final".to_string(), GraphData::int(a + b + 1));
            outputs
        },
        Some("MergeNode"),
        vec![
            (branch_a_id, "result_a", "from_a"),
            (branch_b_id, "result_b", "from_b"),
        ],
        Some(vec![("final", "output")])
    );
    
    let dag = graph.build();
    
    print_section("Mermaid Diagram");
    println!("{}", dag.to_mermaid());
    
    print_section("Sequential Execution (parallel=false)");
    
    let bench = Benchmark::start("Sequential execution");
    let _result_seq = dag.execute_detailed(false, None);
    let bench_result_seq = bench.finish_and_print();
    
    print_section("Parallel Execution (parallel=true)");
    
    let bench = Benchmark::start("Parallel execution");
    let result_par = dag.execute_detailed(true, Some(4));
    let bench_result_par = bench.finish_and_print();
    
    print_section("Results");
    
    println!("📊 Accessing different output levels:\n");
    
    println!("Sequential execution:");
    println!("  Time: {:.3}ms", bench_result_seq.duration_ms);
    
    println!("\nParallel execution:");
    println!("  Time: {:.3}ms", bench_result_par.duration_ms);
    
    // Final context outputs
    println!("\n1. Final context outputs:");
    if let Some(output) = result_par.context.get("output") {
        if let Some(value) = output.as_int() {
            println!("   output: {}", value);
        }
    }
    
    // Node outputs
    println!("\n2. Individual node outputs:");
    println!("   Total nodes executed: {}", result_par.node_outputs.len());
    for (node_id, outputs) in result_par.node_outputs.iter() {
        println!("   Node {}: {} outputs", node_id, outputs.len());
    }
    
    // Branch outputs
    println!("\n3. Branch-specific outputs:");
    println!("   Total branches: {}", result_par.branch_outputs.len());
    for (branch_id, outputs) in result_par.branch_outputs.iter() {
        println!("   Branch {}:", branch_id);
        for (key, value) in outputs.iter() {
            if let Some(v) = value.as_int() {
                println!("     {}: {}", key, v);
            }
        }
    }
    
    println!("\n✅ Successfully accessed all output levels!");
    
    println!();
}
